feat: forward webhooks to codecov

joseph-sentry opened the pull request:
Forwarding webhooks for Sentry running in Monolith mode is considered
out of scope for now.

We want to forward webhooks Sentry receives to Codecov since Codecov
depends on receiving webhooks from the code forges it supports.

We're starting with only GitHub webhooks for now.

We're using the existing WebhookPayload system to forward webhooks to
Codecov by creating an additional WebhookPayload object in the
GithubRequestParser. Codecov will do additional filtering on its end
to filter out irrelevant webhooks so we don't have to do any filtering
on Sentry's end, this is to avoid keeping track of what's relevant to
Codecov in Sentry.

I'm adding a destination_type column to the WebhookPayload so we can
distinguish between payloads meant for Codecov and other Sentry
regions because when we make the request in the drain_mailbox we want
to run different logic, using the CodecovApiClient.

I also made the choice to use a different mailbox_name so that the
forwarding doesn't interfere with the webhooks being sent to the
region silos.

When this is fully rolled out we will be forwarding all webhooks to
Codecov. We will roll this feature out by using the external ID (the
installation ID contained in the webhook) and an Option to determine
the rollout rate as it was done in https://github.com/getsentry/sentry/pull/64395.

I have code that emits metrics so I can observe if my assumptions
about the installation ID are broken.

I also modify the CodecovApiClient to not require the git_provider_org
since webhooks are orthogonal to orgs. The Codecov code that validates
the JWT for the webhooks endpoint doesn't check any claims.


Depends on: https://github.com/getsentry/sentry/pull/94473

markstory commented on src/sentry/hybridcloud/tasks/deliver_webhooks.py:
```diff
@@ -521,3 +540,72 @@ def perform_request(payload: WebhookPayload) -> None:
+
+def perform_codecov_request(payload: WebhookPayload) -> None:
+    """
+    We're dont retrying Codecov forwarding requests for now. We want to prove out that it would work.
```
```suggestion
    We don't retrying Codecov forwarding requests for now. We want to prove out that it would work.
```

markstory commented on src/sentry/integrations/middleware/hybrid_cloud/parser.py:
```diff
@@ -337,3 +338,24 @@ def get_regions_from_organizations(
+            request=self.request,
+        )
+
+        return HttpResponse(status=status.HTTP_202_ACCEPTED)
```
You don't ever use this response. Should this method return None instead?

joseph-sentry commented on src/sentry/hybridcloud/tasks/deliver_webhooks.py:
```diff
@@ -521,3 +540,78 @@ def perform_request(payload: WebhookPayload) -> None:
+                    extra={**logging_context},
+                )
+
+        except ConfigurationError as err:
+            metrics.incr(
+                "hybridcloud.deliver_webhooks.send_request_to_codecov.configuration_error",
+            )
+            logger.warning(
+                "deliver_webhooks.send_request_to_codecov.configuration_error",
+                extra={"error": str(err), **logging_context},
+            )
+        except orjson.JSONDecodeError as err:
+            metrics.incr(
+                "hybridcloud.deliver_webhooks.send_request_to_codecov.json_decode_error",
+            )
+            logger.warning(
+                "deliver_webhooks.send_request_to_codecov.json_decode_error",
+                extra={"error": str(err), **logging_context},
+            )
```
Oh, I don't want it to retry for now, i'd like for it to fail silently to start. My rationale is that I don't want to ship something broken and have it do 3(?) times the work. I want to validate that what I've shipped is not inherently broken, then add retries after, since I'll know that the retries are more likely due to ephemeral issues rather than a result of my unvalidated code.

markstory reviewed (approved):
Looks good to me.

GabeVillalobos commented on src/sentry/middleware/integrations/parsers/github.py:
```diff
@@ -52,6 +52,8 @@ def get_response(self):
         except orjson.JSONDecodeError:
             return HttpResponse(status=400)
 
+        self.forward_to_codecov(external_id=self._get_external_id(event=event))
```
Could we wrap this in a try/except just in case this method raises?

wedamija reviewed (commented):
This is reasonably complicated, would you mind splitting out the db changes into a separate pr so that they can be deployed separately?

joseph-sentry committed:
feat: forward webhooks to codecov

Forwarding webhooks for Sentry running in Monolith mode is considered
out of scope for now.

We want to forward webhooks Sentry receives to Codecov since Codecov
depends on receiving webhooks from the code forges it supports.

We're starting with only GitHub webhooks for now.

We're using the existing WebhookPayload system to forward webhooks to
Codecov by creating an additional WebhookPayload object in the
GithubRequestParser. Codecov will do additional filtering on its end
to filter out irrelevant webhooks so we don't have to do any filtering
on Sentry's end, this is to avoid keeping track of what's relevant to
Codecov in Sentry.

I also made the choice to use a different mailbox_name so that the
forwarding doesn't interfere with the webhooks being sent to the
region silos.

When this is fully rolled out we will be forwarding all webhooks to
Codecov. We will roll this feature out by using the external ID (the
installation ID contained in the webhook) and an Option to determine
the rollout rate as it was done in #64395.

I have code that emits metrics so I can observe if my assumptions
about the installation ID are broken.

I also modify the CodecovApiClient to not require the git_provider_org
since webhooks are orthogonal to orgs. The Codecov code that validates
the JWT for the webhooks endpoint doesn't check any claims.

joseph-sentry commented:
> This is reasonably complicated, would you mind splitting out the db changes into a separate pr so that they can be deployed separately?

@wedamija I split it out into https://github.com/getsentry/sentry/pull/93855

Jeffreyhung reviewed (changes requested):
Putting a pause on this, checking on my side to make sure this change will not violate any DPA or BAA that we have committed to customers

joseph-sentry committed:
feat: filter codecov forwarded webhooks according to options

we want to be able to filter forwarded webhooks to Codecov according to the
Sentry regions that a webhook would be forwarded to, if it's meant for a Sentry
region, i.e. it's not going to be handled in the control silo.

so i'm adding an option and in the github middleware we check what Sentry
regions a webhook is destined for, against the values in the option. If there's
any overlap, we forward the webhook to Codecov as well. If not, that webhook
will not be forwarded to Codecov.

joseph-sentry committed:
Merge branch 'master' into joseph/webhooks

joseph-sentry committed:
type: replace any with actual types in codecov client

joseph-sentry committed:
chore: undo unnecessary change to codecov base url default

joseph-sentry committed:
fix type

leeandher commented on src/sentry/testutils/outbox.py:
```diff
@@ -65,19 +65,21 @@ def assert_webhook_payloads_for_mailbox(
     region_names_set = set(region_names)
     messages = WebhookPayload.objects.filter(mailbox_name=mailbox_name)
-    message_count = messages.count()
+    message_count = messages.filter(region_name__isnull=False).count()
```
I'd rename this var since the contents has a new constraint

leeandher commented on src/sentry/testutils/outbox.py:
```diff
@@ -65,19 +65,21 @@ def assert_webhook_payloads_for_mailbox(
     :param mailbox_name: The mailbox name that messages should be found in.
-    :param region_names: The regions each messages should be queued for
+    :param region_names: Optional list of regions each messages should be queued for
+    :param destination_type: The destination type of the messages
```
I think it's a mapping of destination type to the message count for the desintation? Probably worth including in this docstring.

leeandher commented on src/sentry/testutils/outbox.py:
```diff
@@ -89,12 +91,28 @@ def assert_webhook_payloads_for_mailbox(
         raise Exception(f"WebhookPayload not found for some region_names: {str(region_names_set)}")
+
+    if destination_types and len(destination_types) != 0:
+        raise Exception(
+            f"WebhookPayload not found for some destination_types: {str(destination_types)}"
+        )
```
I think this error can be improved, It's actually that not enought webhookPayloads were found for a given destination type right, not that they weren't found at all
```suggestion
        exc_strs = [f"Missing {destination_count} WebhookPayloads for {destination_type} destination. \n" for [destination_type, destination_count] in destination_types]
        raise Exception(''.join(exc_strs))
```

best i could do in github's markdown editor lol, feel free to do your own thing

leeandher reviewed (approved):
small fixes, but overall looks good to me 🚀

joseph-sentry commented on src/sentry/integrations/middleware/hybrid_cloud/parser.py:
```diff
@@ -349,3 +352,42 @@ def get_regions_from_organizations(
+            return
+
+        if not external_id:
+            metrics.incr("codecov.forward-webhooks.missing-external", sample_rate=0.01)
```
I think i'm just going to have these ones use the default sample rate

joseph-sentry committed:
address feedback

joseph-sentry committed:
return after handling exception in deliver_webhooks

joseph-sentry committed:
fix tests

Jeffreyhung reviewed (approved):
LGTM!
